libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1282,35 +1299,53 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 
 			if (hasSelectorOrSignature)
 			{
-				// stack: <selector> <memory pointer>
-				solAssert(arguments.size() >= 1, "");
-				Type const* selectorType = arguments[0]->annotation().type;
-				utils().moveIntoStack(selectorType->sizeOnStack());
-				Type const* dataOnStack = selectorType;
-				// stack: <memory pointer> <selector>
-				if (function.kind() == FunctionType::Kind::ABIEncodeWithSignature)
+				Type const* dataOnStack = nullptr;
+
+				if (function.kind() == FunctionType::Kind::ABIEncodeCall)
 				{
-					// hash the signature
-					if (auto const* stringType = dynamic_cast<StringLiteralType const*>(selectorType))
-					{
-						m_context << util::selectorFromSignature(stringType->value());
-						dataOnStack = TypeProvider::fixedBytes(4);
-					}
+					auto const functionPtr = dynamic_cast<FunctionTypePointer>(arguments[0]->annotation().type);
+					solAssert(functionPtr, "");
+
+					if (functionPtr->hasDeclaration())
```

Comment:
Does the `accept()` call above not put the funciton pointer on the stack as a value? I.e. I think you can remove the "if" part and only keep the "else" case.